cpp/src/arrow/python/pandas_convert.cc

```diff
@@ -80,6 +80,19 @@ static inline bool PyObject_is_string(const PyObject* obj) {
 #endif
 }
 
+static inline bool PyObject_is_float(const PyObject* obj) {
+  return PyFloat_Check(obj);
+}
+
+static inline bool PyObject_is_integer(const PyObject* obj) {
+#if PY_MAJOR_VERSION >= 3
+  return (!PyBool_Check(obj)) && (PyLong_Check(obj) || PyArray_IsIntegerScalar(obj));
```

Comment:
`PyArray_IsIntegerScalar` uses  `PyInt_Check` with Python 3.4 and `numpy==1.9.x` Thus we probably need to raise our minimal NumPy version to `>=1.10`.  For the manylinux1 build, you have to adjust it in `python/manylinux1/scripts/build_virtualenvs.sh`